src/types/resolveDescriptors.ts

```diff
@@ -841,12 +841,19 @@ export function resolveDescriptors(ctx: CompilerContext, Ast: FactoryAst) {
 
         // Check parameter names
         const exNames: Set<string> = new Set();
-        for (const param of params) {
+        for (const [i, param] of params.entries()) {
             if (isSelfId(param.name)) {
-                throwCompilationError(
-                    'Parameter name "self" is reserved',
-                    param.loc,
-                );
+                if (i === 0) {
+                    throwCompilationError(
+                        'Parameter name "self" is reserved for functions with "extends" modifier',
+                        param.loc,
+                    );
```

Comment:
why do we check this at each iteration? the proper way to do it is to split the head and the tail of the parameters list and process those separately

Reply:
have a look please